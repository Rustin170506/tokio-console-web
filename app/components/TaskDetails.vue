<template>
    <div>
        <div v-if="pending" class="flex items-center space-x-4">
            <USkeleton class="h-12 w-12" :ui="{ rounded: 'rounded-full' }" />
            <div class="space-y-2">
                <USkeleton class="h-4 w-[250px]" />
                <USkeleton class="h-4 w-[200px]" />
                <USkeleton class="h-4 w-[150px]" />
                <USkeleton class="h-4 w-[100px]" />
                <USkeleton class="h-4 w-[50px]" />
            </div>
        </div>

        <div v-else>
            <UCard
                :ui="{
                    body: {
                        base: 'xl:flex',
                    },
                }"
            >
                <div class="space-y-2 flex-1 mr-4">
                    <UDivider type="dotted">
                        <div
                            class="flex items-center text-black dark:text-white font-bold text-md"
                        >
                            <span>Task</span>
                            <UIcon :name="taskBasicInfo.state" dynamic />
                        </div>
                    </UDivider>
                    <InfoField name="ID" :value="taskBasicInfo.idString" />
                    <InfoField name="Name" :value="taskBasicInfo.name" />
                    <InfoField
                        name="Location"
                        :value="taskBasicInfo.location"
                    />
                    <InfoField name="Total" :value="taskBasicInfo.total" />
                    <InfoField name="Busy">
                        <div class="flex">
                            <span :class="taskBasicInfo.busy.class">{{
                                taskBasicInfo.busy.value
                            }}</span>
                            <span class="mx-2">
                                ({{ taskBasicInfo.busyPercentage }})
                            </span>
                        </div>
                    </InfoField>
                    <InfoField name="Scheduled">
                        <div class="flex">
                            <span :class="taskBasicInfo.sched.class">{{
                                taskBasicInfo.sched.value
                            }}</span>
                            <span class="mx-2">
                                ({{ taskBasicInfo.scheduledPercentage }})
                            </span>
                        </div>
                    </InfoField>
                    <InfoField name="Idle">
                        <div class="flex">
                            <span :class="taskBasicInfo.idle.class">{{
                                taskBasicInfo.idle.value
                            }}</span>
                            <span class="mx-2">
                                ({{ taskBasicInfo.idlePercentage }})
                            </span>
                        </div>
                    </InfoField>
                </div>

                <UDivider
                    class="w-px mx-2"
                    color="gray"
                    orientation="vertical"
                />

                <div class="space-y-2 xl:ml-4 flex-1">
                    <UDivider
                        label="Waker"
                        type="dotted"
                        :ui="{
                            label: 'text-black dark:text-white font-bold text-md',
                        }"
                    />
                    <InfoField name="Current wakers">
                        <div class="flex">
                            <span class="mr-1">{{
                                taskBasicInfo.wakerCount
                            }}</span>
                            (
                            <InfoField
                                name="Clones"
                                :value="taskBasicInfo.wakerClones"
                            />
                            <span class="mr-2">, </span>
                            <InfoField
                                name="Drops"
                                :value="taskBasicInfo.wakerDrops"
                            />
                            )
                        </div>
                    </InfoField>
                    <InfoField name="Woken">
                        <div class="flex">
                            <span class="mr-2">{{
                                `${taskBasicInfo.wakes} times,`
                            }}</span>
                            <InfoField
                                v-if="taskBasicInfo.lastWokenDuration"
                                name="Last woken"
                            >
                                <span
                                    :class="
                                        taskBasicInfo.lastWokenDuration.class
                                    "
                                >
                                    {{
                                        taskBasicInfo.lastWokenDuration.toString()
                                    }}
                                </span>
                                <span class="mx-1">ago</span>
                            </InfoField>
                        </div>
                    </InfoField>
                    <UDivider
                        label="Fields"
                        type="dotted"
                        :ui="{
                            label: 'text-black dark:text-white font-bold text-md',
                        }"
                    />
                    <div
                        v-for="(item, index) in taskBasicInfo.fields"
                        :key="index"
                        class="mb-1"
                    >
                        <p>
                            <span class="text-blue-600 dark:text-blue-400">{{
                                item.name
                            }}</span>
                            <span class="text-gray-500 dark:text-gray-400"
                                >=</span
                            >
                            <span class="text-green-600 dark:text-green-400">{{
                                item.value.value
                            }}</span>
                        </p>
                    </div>
                </div>
            </UCard>
            <UCard class="mt-4" :ui="{ body: { base: 'xl:flex' } }">
                <div class="space-y-2 flex-1 mr-4">
                    <div class="mt-2">
                        <div class="flex justify-center">
                            <InfoField
                                name="Min"
                                :value="taskDetailsInfo.pollTimes.min"
                                class="mr-4"
                            />
                            <InfoField
                                v-for="percentile in taskDetailsInfo.pollTimes.percentiles.slice(
                                    0,
                                    3,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                        </div>
                        <div class="flex justify-center">
                            <InfoField
                                v-for="percentile in taskDetailsInfo.pollTimes.percentiles.slice(
                                    -4,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                            <InfoField
                                name="Max"
                                :value="taskDetailsInfo.pollTimes.max"
                                class="mr-4"
                            />
                        </div>
                    </div>
                    <div class="space-y-2 h-80 w-11/12">
                        <HistogramChart :data="pollTimesHistogramData" />
                    </div>
                </div>

                <UDivider
                    v-if="taskDetailsInfo.scheduledTimes"
                    class="w-px mx-2"
                    color="gray"
                    orientation="vertical"
                />

                <div
                    v-if="taskDetailsInfo.scheduledTimes"
                    class="space-y-2 xl:ml-4 flex-1"
                >
                    <div class="mt-2">
                        <div class="flex justify-center">
                            <InfoField
                                name="Min"
                                :value="taskDetailsInfo.scheduledTimes.min"
                                class="mr-4"
                            />
                            <InfoField
                                v-for="percentile in taskDetailsInfo.scheduledTimes.percentiles.slice(
                                    0,
                                    3,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                        </div>
                        <div class="flex justify-center">
                            <InfoField
                                v-for="percentile in taskDetailsInfo.scheduledTimes.percentiles.slice(
                                    -4,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                            <InfoField
                                name="Max"
                                :value="taskDetailsInfo.scheduledTimes.max"
                                class="mr-4"
                            />
                        </div>
                    </div>
                    <div class="space-y-2 h-80 w-11/12">
                        <HistogramChart :data="scheduledTimesHistogramData" />
                    </div>
                </div>
            </UCard>
        </div>
    </div>
</template>
<script setup lang="ts">
import { type TimesDetails, toTaskDetails } from "~/types/timesDetails";
import { toTaskBasicInfo } from "~/types/taskBasicInfo";

const router = useRouter();
const route = useRoute();

const { pending, task, taskDetails, lastUpdatedAt, closed } = useTaskDetails(
    BigInt(route.params.id as string),
);

// If the task is not available, redirect to the home page.
if (!task || !lastUpdatedAt.value) {
    router.push("/");
}

const taskBasicInfo = computed(() =>
    toTaskBasicInfo(task!, lastUpdatedAt.value!),
);

const taskDetailsInfo = computed(() => toTaskDetails(taskDetails.value));

const buildHistogramData = (times: TimesDetails | undefined, label: string) =>
    times
        ? {
              labels: times.histogram.map((h) => h.duration.toString()),
              datasets: [
                  {
                      label,
                      data: times.histogram.map((h) => h.count),
                  },
              ],
          }
        : undefined;

const pollTimesHistogramData = computed(() =>
    buildHistogramData(taskDetailsInfo.value.pollTimes, "Poll Times"),
);

const scheduledTimesHistogramData = computed(() =>
    buildHistogramData(taskDetailsInfo.value.scheduledTimes, "Scheduled Times"),
);

onBeforeUnmount(() => {
    closed.value = true;
});
</script>
