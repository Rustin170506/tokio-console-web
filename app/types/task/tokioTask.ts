import { Duration, Timestamp } from "../common/duration";
import type { Field } from "../common/field";
import { Warning, type Linter, type Warn } from "../warning/warn";
import type { TokioTaskStats } from "./tokioTaskStats";

export enum TaskState {
    Completed,
    Running,
    Scheduled,
    Idle,
}

export enum TaskLintResult {
    Linted,
    RequiresRecheck,
}

export class TokioTask {
    // The task's pretty (console-generated, sequential) task ID.
    //
    // This is NOT the `tracing::span::Id` for the task's tracing span on the
    // remote.
    id: bigint;
    // The `tokio::task::Id` in the remote tokio runtime.
    taskId?: bigint;
    // The `tracing::span::Id` on the remote process for this task's span.
    //
    // This is used when requesting a task details stream
    spanId: bigint;
    // A precomputed short description string used in the async ops table
    shortDesc: string;
    // Fields that don't have their own column, pre-formatted
    fields: Array<Field>;
    // The task statistics that are updated over the lifetime of the task.
    stats: TokioTaskStats;
    // The target of the span representing the task.
    target: string;
    // The name of the task
    name?: string;
    // The source file and line number the task was spawned from
    location: string;
    // The kind of task, currently one of task, blocking, block_on, local
    kind: string;
    // Currently active warnings for this task.
    /* eslint-disable no-use-before-define */
    warnings: Array<Warn<TokioTask>>;

    constructor(
        id: bigint,
        taskId: bigint | undefined,
        spanId: bigint,
        shortDesc: string,
        formattedFields: Array<Field>,
        stats: TokioTaskStats,
        target: string,
        name: string | undefined,
        location: string,
        kind: string,
    ) {
        this.id = id;
        this.taskId = taskId;
        this.spanId = spanId;
        this.shortDesc = shortDesc;
        this.fields = formattedFields;
        this.stats = stats;
        this.target = target;
        this.name = name;
        this.location = location;
        this.kind = kind;
        this.warnings = [];
    }

    totalDuration(since: Timestamp): Duration {
        const duration = since.subtract(this.stats.createdAt);
        return this.stats.total ?? duration;
    }

    busyDuration(since: Timestamp): Duration {
        if (this.stats.lastPollStarted) {
            if (
                !this.stats.lastPollEnded ||
                this.stats.lastPollStarted > this.stats.lastPollEnded
            ) {
                // In this case the task is being polled at the moment.
                const currentTimeInPollDuration = since.subtract(
                    this.stats.lastPollStarted,
                );
                return this.stats.busy.add(currentTimeInPollDuration);
            }
        }
        return this.stats.busy;
    }

    scheduledDuration(since: Timestamp): Duration {
        if (this.stats.lastWake) {
            if (
                !this.stats.lastPollStarted ||
                this.stats.lastWake > this.stats.lastPollStarted
            ) {
                // In this case the task is scheduled, but has not yet been polled.
                const currentTimeSinceWakeDuration = since.subtract(
                    this.stats.lastWake,
                );
                return this.stats.scheduled.add(currentTimeSinceWakeDuration);
            }
        }
        return this.stats.scheduled;
    }

    idleDuration(since: Timestamp): Duration {
        if (this.stats.idle) {
            return this.stats.idle;
        } else {
            const total = this.totalDuration(since);
            const busy = this.busyDuration(since);
            const scheduled = this.scheduledDuration(since);
            if (total.greaterThan(busy.add(scheduled))) {
                return total.subtract(busy.add(scheduled));
            }
        }
        return new Duration(BigInt(0), 0);
    }

    isRunning(): boolean {
        if (this.stats.lastPollStarted && this.stats.lastPollEnded) {
            return this.stats.lastPollStarted.greaterThan(
                this.stats.lastPollEnded,
            );
        }
        if (
            this.stats.lastPollStarted &&
            this.stats.lastPollEnded === undefined
        ) {
            return true;
        }
        return false;
    }

    isScheduled(): boolean {
        if (this.stats.lastWake && this.stats.lastPollStarted) {
            return this.stats.lastWake.greaterThan(this.stats.lastPollStarted);
        }
        return false;
    }

    isCompleted(): boolean {
        return this.stats.total !== undefined;
    }

    isBlocking(): boolean {
        return ["block_on", "blocking"].includes(this.kind);
    }

    // Returns whether this task has signaled via its waker to run again.
    // Once the task has been polled, this is changed back to false.
    isAwakened(): boolean {
        // Before the first poll, the task is waiting on the executor to run it
        // for the first time.
        if (this.totalPolls() === 0n) {
            return true;
        }

        if (this.stats.lastWake && this.stats.lastPollStarted) {
            return this.stats.lastWake.greaterThan(this.stats.lastPollStarted);
        }

        if (this.stats.lastWake && this.stats.lastPollStarted === undefined) {
            return true;
        }

        return false;
    }

    totalPolls(): bigint {
        return this.stats.polls;
    }

    state(): TaskState {
        if (this.isCompleted()) {
            return TaskState.Completed;
        }

        if (this.isRunning()) {
            return TaskState.Running;
        }

        if (this.isScheduled()) {
            return TaskState.Scheduled;
        }

        return TaskState.Idle;
    }

    wakerCount(): bigint {
        return this.stats.wakerClones - this.stats.wakerDrops > 0n
            ? this.stats.wakerClones - this.stats.wakerDrops
            : 0n;
    }

    /**
     * Returns the elapsed time since the task was last woken, relative to
     * given `now` timestamp.
     *
     * Returns `undefined` if the task has never been woken, or if it was last woken
     * more recently than `now` (which *shouldn't* happen as long as `now` is the
     * timestamp of the last stats update...)
     */
    sinceWakeDuration(now: Timestamp): Duration | undefined {
        if (this.stats.lastWake) {
            return now.subtract(this.stats.lastWake);
        }
    }

    durationPercent(now: Timestamp, amt: Duration): number {
        const total = this.totalDuration(now);
        let percent = (amt.asSeconds() / total.asSeconds()) * 100;
        if (percent > 100) {
            percent = 100;
        }

        return percent;
    }

    lint(linters: Array<Linter<TokioTask>>): TaskLintResult {
        this.warnings.length = 0;
        let recheck = false;
        linters.forEach((linter) => {
            const result = linter.check(this);
            switch (result) {
                case Warning.Ok:
                    break;
                case Warning.Recheck:
                    recheck = true;
                    break;
                case Warning.Warn:
                    this.warnings.push(linter);
                    break;
            }
        });
        if (recheck) {
            return TaskLintResult.RequiresRecheck;
        } else {
            return TaskLintResult.Linted;
        }
    }
}
