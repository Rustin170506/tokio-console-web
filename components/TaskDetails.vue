<template>
    <div>
        <div v-if="pending" class="flex items-center space-x-4">
            <USkeleton class="h-12 w-12" :ui="{ rounded: 'rounded-full' }" />
            <div class="space-y-2">
                <USkeleton class="h-4 w-[250px]" />
                <USkeleton class="h-4 w-[200px]" />
            </div>
        </div>

        <div v-else>
            <UCard
                :ui="{
                    body: {
                        base: 'xl:flex',
                    },
                }"
            >
                <div class="space-y-2 flex-1">
                    <UDivider type="dotted">
                        <div
                            class="flex items-center text-black dark:text-white font-bold text-md"
                        >
                            <span>Task</span>
                            <UIcon :name="taskBasicInfo.state" />
                        </div>
                    </UDivider>
                    <TaskInfoField name="ID" :value="taskBasicInfo.idString" />
                    <TaskInfoField name="Name" :value="taskBasicInfo.name" />
                    <TaskInfoField
                        name="Location"
                        :value="taskBasicInfo.location"
                    />
                    <TaskInfoField name="Total" :value="taskBasicInfo.total" />
                    <TaskInfoField name="Busy">
                        <div class="flex">
                            <span :class="taskBasicInfo.busy.class">{{
                                taskBasicInfo.busy.value
                            }}</span>
                            <span class="mx-2">
                                ({{ taskBasicInfo.busyPercentage }})
                            </span>
                        </div>
                    </TaskInfoField>
                    <TaskInfoField name="Scheduled">
                        <div class="flex">
                            <span :class="taskBasicInfo.sched.class">{{
                                taskBasicInfo.sched.value
                            }}</span>
                            <span class="mx-2">
                                ({{ taskBasicInfo.scheduledPercentage }})
                            </span>
                        </div>
                    </TaskInfoField>
                    <TaskInfoField name="Idle">
                        <div class="flex">
                            <span :class="taskBasicInfo.idle.class">{{
                                taskBasicInfo.idle.value
                            }}</span>
                            <span class="mx-2">
                                ({{ taskBasicInfo.idlePercentage }})
                            </span>
                        </div>
                    </TaskInfoField>
                </div>

                <UDivider
                    class="w-px mx-2"
                    color="gray"
                    orientation="vertical"
                />

                <div class="space-y-2 xl:ml-4 flex-1">
                    <UDivider
                        label="Waker"
                        type="dotted"
                        :ui="{
                            label: 'text-black dark:text-white font-bold text-md',
                        }"
                    />
                    <TaskInfoField name="Current wakers">
                        <div class="flex">
                            <span class="mr-1">{{
                                taskBasicInfo.wakerCount
                            }}</span>
                            (
                            <TaskInfoField
                                name="Clones"
                                :value="taskBasicInfo.wakerClones"
                            />
                            <span class="mr-2">, </span>
                            <TaskInfoField
                                name="Drops"
                                :value="taskBasicInfo.wakerDrops"
                            />
                            )
                        </div>
                    </TaskInfoField>
                    <TaskInfoField name="Woken">
                        <div class="flex">
                            <span class="mr-2">{{
                                `${taskBasicInfo.wakes} times,`
                            }}</span>
                            <TaskInfoField
                                v-if="
                                    taskBasicInfo.lastWokenDuration !==
                                    undefined
                                "
                                name="Last woken"
                            >
                                <span
                                    :class="
                                        taskBasicInfo.lastWokenDuration!.class
                                    "
                                >
                                    {{
                                        taskBasicInfo.lastWokenDuration!.toString()
                                    }}
                                </span>
                                <span class="mx-1">ago</span>
                            </TaskInfoField>
                        </div>
                    </TaskInfoField>
                    <UDivider
                        label="Fields"
                        type="dotted"
                        :ui="{
                            label: 'text-black dark:text-white font-bold text-md',
                        }"
                    />
                    <div
                        v-for="(item, index) in taskBasicInfo.fields"
                        :key="index"
                        class="mb-1"
                    >
                        <p>
                            <span class="text-blue-600 dark:text-blue-400">{{
                                item.name
                            }}</span>
                            <span class="text-gray-500 dark:text-gray-400"
                                >=</span
                            >
                            <span class="text-green-600 dark:text-green-400">{{
                                item.value
                            }}</span>
                        </p>
                    </div>
                </div>
            </UCard>
            <UCard class="mt-4" :ui="{ body: { base: 'xl:flex' } }">
                <div class="space-y-2 flex-1">
                    <div class="mt-2">
                        <div class="flex justify-center">
                            <TaskInfoField
                                name="Min"
                                :value="taskDetailsInfo.pollTimes.min"
                                class="mr-4"
                            />
                            <TaskInfoField
                                v-for="percentile in taskDetailsInfo.pollTimes.percentiles.slice(
                                    0,
                                    3,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                        </div>
                        <div class="flex justify-center">
                            <TaskInfoField
                                v-for="percentile in taskDetailsInfo.pollTimes.percentiles.slice(
                                    -4,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                            <TaskInfoField
                                name="Max"
                                :value="taskDetailsInfo.pollTimes.max"
                                class="mr-4"
                            />
                        </div>
                    </div>
                    <div class="space-y-2 h-80 w-11/12">
                        <HistogramChart :data="pollTimesHistogramData" />
                    </div>
                </div>

                <UDivider
                    v-if="taskDetailsInfo.scheduledTimes !== undefined"
                    class="w-px mx-2"
                    color="gray"
                    orientation="vertical"
                />

                <div
                    v-if="taskDetailsInfo.scheduledTimes !== undefined"
                    class="space-y-2 xl:ml-4 flex-1"
                >
                    <div class="mt-2">
                        <div class="flex justify-center">
                            <TaskInfoField
                                name="Min"
                                :value="taskDetailsInfo.scheduledTimes.min"
                                class="mr-4"
                            />
                            <TaskInfoField
                                v-for="percentile in taskDetailsInfo.scheduledTimes.percentiles.slice(
                                    0,
                                    3,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                        </div>
                        <div class="flex justify-center">
                            <TaskInfoField
                                v-for="percentile in taskDetailsInfo.scheduledTimes.percentiles.slice(
                                    -4,
                                )"
                                :key="percentile.percentile"
                                :name="percentile.percentile"
                                :value="percentile.duration"
                                class="mr-4"
                            />
                            <TaskInfoField
                                name="Max"
                                :value="taskDetailsInfo.scheduledTimes.max"
                                class="mr-4"
                            />
                        </div>
                    </div>
                    <div class="space-y-2 h-80 w-11/12">
                        <HistogramChart :data="scheduledTimesHistogramData" />
                    </div>
                </div>
            </UCard>
        </div>
    </div>
</template>
<script setup lang="ts">
const router = useRouter();
const route = useRoute();

const { pending, task, taskDetails } = useTaskDetails(
    BigInt(route.params.id as string),
);

// If the task is undefined, redirect to the home page.
if (task === undefined) {
    router.push("/");
}

const taskBasicInfo = computed(() => {
    return toTaskBasicInfo(task!);
});

const taskDetailsInfo = computed(() => {
    return toTaskDetails(taskDetails.value);
});

function createHistogramData(times: TimesDetails | undefined, label: string) {
    if (times === undefined) {
        return undefined;
    }

    return {
        labels: times.histogram.map((h) => h.duration.toString()),
        datasets: [
            {
                label,
                data: times.histogram.map((h) => h.count),
            },
        ],
    };
}

const pollTimesHistogramData = computed(() => {
    return createHistogramData(taskDetailsInfo.value.pollTimes, "Poll Times");
});

const scheduledTimesHistogramData = computed(() => {
    return createHistogramData(
        taskDetailsInfo.value.scheduledTimes,
        "Scheduled Times",
    );
});
</script>
