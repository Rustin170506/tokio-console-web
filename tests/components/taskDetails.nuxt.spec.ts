import { expect, it } from "vitest";
import { mockNuxtImport, mountSuspended } from "nuxt-vitest/utils";
import TaskDetails from "~/components/TaskDetails.vue";
import { Duration, Timestamp } from "~/types/common/duration";
import { TokioTask } from "~/types/task/tokioTask";
import type { TokioTaskDetails } from "~/types/task/tokioTaskDetails";

mockNuxtImport("useRoute", () => {
    return () => {
        return {
            params: {
                id: "1",
            },
        };
    };
});

mockNuxtImport("useTaskDetails", () => {
    return (_id: bigint) => {
        const pending = ref<boolean>(false);
        const formattedFields = [{ name: "target", value: "tokio:task" }];
        const stats = {
            polls: 100n,
            createdAt: new Duration(1000n, 0),
            busy: new Duration(500n, 0),
            scheduled: new Duration(300n, 0),
            lastPollStarted: new Duration(1000n, 0),
            lastPollEnded: new Timestamp(0n, 0),
            idle: new Duration(200n, 0),
            wakes: 1n,
            wakerClones: 199n,
            wakerDrops: 198n,
            selfWakes: 100n,
            lastWake: new Duration(1000n, 0),
        };
        const task = new TokioTask(
            2n,
            1n,
            1n,
            "some task",
            formattedFields,
            stats,
            "tokio:task",
            "task1",
            "app.rs/68:10",
            "task",
        );
        const taskDetails = ref<TokioTaskDetails>({
            pollTimes: {
                percentiles: [
                    { percentile: 0.1, duration: new Duration(100n, 0) },
                    { percentile: 0.25, duration: new Duration(200n, 0) },
                    { percentile: 0.5, duration: new Duration(300n, 0) },
                    { percentile: 0.75, duration: new Duration(400n, 0) },
                    { percentile: 0.9, duration: new Duration(700n, 0) },
                    { percentile: 0.95, duration: new Duration(800n, 0) },
                    { percentile: 0.99, duration: new Duration(990n, 0) },
                ],
                histogram: [
                    { duration: new Duration(100n, 0), count: 10n },
                    { duration: new Duration(200n, 0), count: 10n },
                    { duration: new Duration(300n, 0), count: 20n },
                    { duration: new Duration(400n, 0), count: 30n },
                    { duration: new Duration(500n, 0), count: 51n },
                    { duration: new Duration(600n, 0), count: 0n },
                    { duration: new Duration(700n, 0), count: 0n },
                    { duration: new Duration(800n, 0), count: 0n },
                    { duration: new Duration(900n, 0), count: 0n },
                    { duration: new Duration(1000n, 0), count: 1n },
                ],
                min: new Duration(100n, 0),
                max: new Duration(1000n, 0),
            },
            scheduledTimes: {
                percentiles: [
                    { percentile: 0.1, duration: new Duration(100n, 0) },
                    { percentile: 0.25, duration: new Duration(200n, 0) },
                    { percentile: 0.5, duration: new Duration(300n, 0) },
                    { percentile: 0.75, duration: new Duration(400n, 0) },
                    { percentile: 0.9, duration: new Duration(700n, 0) },
                    { percentile: 0.95, duration: new Duration(800n, 0) },
                    { percentile: 0.99, duration: new Duration(990n, 0) },
                ],
                histogram: [
                    { duration: new Duration(100n, 0), count: 10n },
                    { duration: new Duration(200n, 0), count: 10n },
                    { duration: new Duration(300n, 0), count: 20n },
                    { duration: new Duration(400n, 0), count: 30n },
                    { duration: new Duration(500n, 0), count: 51n },
                    { duration: new Duration(600n, 0), count: 0n },
                    { duration: new Duration(700n, 0), count: 0n },
                    { duration: new Duration(800n, 0), count: 0n },
                    { duration: new Duration(900n, 0), count: 0n },
                    { duration: new Duration(1000n, 0), count: 1n },
                ],
                min: new Duration(100n, 0),
                max: new Duration(1000n, 0),
            },
        });
        const lastUpdatedAt = ref<Timestamp>(new Timestamp(1000n, 0));

        return { pending, task, taskDetails, lastUpdatedAt };
    };
});

it("TaskDetails", async () => {
    const component = await mountSuspended(TaskDetails);
    expect(component.html()).toMatchSnapshot();
});
