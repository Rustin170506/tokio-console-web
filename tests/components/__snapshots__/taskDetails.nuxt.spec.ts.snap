// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskDetails 1`] = `
"<div>
  <div>
    <div class=\\"overflow-hidden rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900\\">
      <!--v-if-->
      <div class=\\"xl:flex px-4 py-5 sm:p-6\\">
        <div class=\\"space-y-2 flex-1\\">
          <div class=\\"flex items-center align-center text-center w-full flex-row\\">
            <div class=\\"flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted\\"></div>
            <div class=\\"font-medium text-gray-700 dark:text-gray-200 flex mx-3 whitespace-nowrap\\"><span class=\\"text-black dark:text-white font-bold text-md\\">Task</span></div>
            <div class=\\"flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted\\"></div>
          </div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">ID</span><span class=\\"mr-1\\">: </span><span class=\\"\\">1</span></div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Name</span><span class=\\"mr-1\\">: </span><span class=\\"\\">task1</span></div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Location</span><span class=\\"mr-1\\">: </span><span class=\\"\\">app.rs/68:10</span></div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Total</span><span class=\\"mr-1\\">: </span><span class=\\"text-green-500 dark:text-green-300\\">16m 40s </span></div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Busy</span><span class=\\"mr-1\\">: </span>
            <div class=\\"flex\\"><span class=\\"text-green-500 dark:text-green-300\\">25m </span><span class=\\"mx-2\\"> (100.00%) </span></div>
          </div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Scheduled</span><span class=\\"mr-1\\">: </span>
            <div class=\\"flex\\"><span class=\\"text-green-500 dark:text-green-300\\">5m </span><span class=\\"mx-2\\"> (30.00%) </span></div>
          </div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Idle</span><span class=\\"mr-1\\">: </span>
            <div class=\\"flex\\"><span class=\\"text-green-500 dark:text-green-300\\">3m 20s </span><span class=\\"mx-2\\"> (20.00%) </span></div>
          </div>
        </div>
        <div class=\\"flex items-center align-center text-center flex-col w-px mx-2\\" color=\\"gray\\">
          <div class=\\"flex border-gray-200 dark:border-gray-800 h-full border-s border-solid\\"></div>
          <!--v-if-->
        </div>
        <div class=\\"space-y-2 xl:ml-4 flex-1\\">
          <div class=\\"flex items-center align-center text-center w-full flex-row\\">
            <div class=\\"flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted\\"></div>
            <div class=\\"font-medium text-gray-700 dark:text-gray-200 flex mx-3 whitespace-nowrap\\"><span class=\\"text-black dark:text-white font-bold text-md\\">Waker</span></div>
            <div class=\\"flex border-gray-200 dark:border-gray-800 w-full border-t border-dotted\\"></div>
          </div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Current wakers</span><span class=\\"mr-1\\">: </span>
            <div class=\\"flex\\"><span class=\\"mr-1\\">1</span> ( <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Clones</span><span class=\\"mr-1\\">: </span><span class=\\"\\">199</span></div><span class=\\"mr-2\\">, </span>
              <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Drops</span><span class=\\"mr-1\\">: </span><span class=\\"\\">198</span></div> )
            </div>
          </div>
          <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Woken</span><span class=\\"mr-1\\">: </span>
            <div class=\\"flex\\"><span class=\\"mr-2\\">1 times,</span>
              <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Last woken</span><span class=\\"mr-1\\">: </span><span class=\\"text-gray-500 dark:text-gray-300\\">0ns</span><span class=\\"mx-1\\">ago</span></div>
            </div>
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
    <div class=\\"overflow-hidden rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900 mt-4\\">
      <!--v-if-->
      <div class=\\"xl:flex px-4 py-5 sm:p-6\\">
        <div class=\\"space-y-2 flex-1\\">
          <div class=\\"flex m-2\\"></div>
          <div class=\\"space-y-2 h-80 w-11/12\\"><canvas role=\\"img\\">
              <p></p>
            </canvas></div>
        </div>
        <div class=\\"flex items-center align-center text-center flex-col w-px mx-2\\" color=\\"gray\\">
          <div class=\\"flex border-gray-200 dark:border-gray-800 h-full border-s border-solid\\"></div>
          <!--v-if-->
        </div>
        <div class=\\"space-y-2 flex-1 xl:ml-4\\">
          <div class=\\"flex m-2\\"></div>
          <div class=\\"space-y-2 h-80 w-11/12\\"><canvas role=\\"img\\">
              <p></p>
            </canvas></div>
        </div>
      </div>
      <!--v-if-->
    </div>
  </div>
</div>"
`;
